### DISCUSSION TITLE
State variables selection

### DISCUSSION THREAD

**Discussion by lucas-mior** [OpenModelica/OpenModelica#12364] [Q&A] [answered]
Is there a way to select which state variables are used or at least give a hint to OM?
Currently I check which variables were chosen as state by reading the "solution of the initialization" -> "states variables" message.

**Comment by casella** [comment, marked as answer]
Sure, you can use the stateSelect attribute when declaring variables, e.g.:
```modelica
model M
  Real M(stateSelect = StateSelect.avoid);
  Real p(stateSelect = StateSelect.prefer);
equation
  M = f(p);
  der(M) = ...;
```
in this case OMC will try to use p as state, which means it will differentiate M = f(p) and automatically come up with
```modelica
model M
  Real M(stateSelect = StateSelect.avoid);
  Real p(stateSelect = StateSelect.prefer);
equation
  M = f(p);
  dM_dp*der(p) = ...;
```
where dM_dp is the partial derivative of f(p) w.r.t. p.
